crates/rspack_plugin_mf/src/container/embed_federation_runtime_plugin.rs

```diff
@@ -0,0 +1,244 @@
+use rspack_hook::{plugin, plugin_hook};
+use rspack_plugin_javascript::{JavascriptModulesRenderStartup, JsPlugin, RenderSource};
+use rspack_sources::{ConcatSource, RawStringSource, SourceExt};
+
+use super::{
+  embed_federation_runtime_module::{
+    EmbedFederationRuntimeModule, EmbedFederationRuntimeModuleOptions,
+  },
+  federation_modules_plugin::{AddFederationRuntimeDependencyHook, FederationModulesPlugin},
+  federation_runtime_dependency::FederationRuntimeDependency,
+};
+
+struct FederationRuntimeDependencyCollector {
+  collected_dependency_ids: Arc<Mutex<HashSet<DependencyId>>>,
+}
+
+#[async_trait::async_trait]
+impl AddFederationRuntimeDependencyHook for FederationRuntimeDependencyCollector {
+  async fn run(&self, dependency: &FederationRuntimeDependency) -> Result<()> {
+    self
+      .collected_dependency_ids
+      .lock()
+      .expect("Failed to lock collected_dependency_ids")
+      .insert(dependency.id);
+    Ok(())
+  }
+}
+
+#[plugin]
+#[derive(Debug)]
+pub struct EmbedFederationRuntimePlugin {
+  collected_dependency_ids: Arc<Mutex<HashSet<DependencyId>>>,
+}
+
+impl EmbedFederationRuntimePlugin {
+  pub fn new() -> Self {
+    Self::new_inner(Arc::new(Mutex::new(HashSet::new())))
+  }
+}
+
+#[plugin_hook(CompilationAdditionalChunkRuntimeRequirements for EmbedFederationRuntimePlugin)]
+async fn additional_chunk_runtime_requirements_tree(
+  &self,
+  compilation: &mut Compilation,
+  chunk_ukey: &ChunkUkey,
+  runtime_requirements: &mut RuntimeGlobals,
+) -> Result<()> {
+  let chunk = compilation.chunk_by_ukey.expect_get(chunk_ukey);
+
+  // Skip build time chunks
+  if chunk.name() == Some("build time chunk") {
+    return Ok(());
+  }
+
+  // Check if chunk needs federation runtime support
+  let has_runtime = chunk.has_runtime(&compilation.chunk_group_by_ukey);
+  let has_entry_modules = compilation
+    .chunk_graph
+    .get_number_of_entry_modules(chunk_ukey)
+    > 0;
+
+  // Federation is enabled for runtime chunks or entry chunks
+  let is_enabled = has_runtime || has_entry_modules;
+
+  if is_enabled {
+    // Add STARTUP requirement
+    runtime_requirements.insert(RuntimeGlobals::STARTUP);
+  }
+
+  Ok(())
+}
+
+#[plugin_hook(CompilationRuntimeRequirementInTree for EmbedFederationRuntimePlugin)]
+async fn runtime_requirement_in_tree(
+  &self,
+  compilation: &mut Compilation,
+  chunk_ukey: &ChunkUkey,
+  _all_runtime_requirements: &RuntimeGlobals,
+  _runtime_requirements: &RuntimeGlobals,
+  _runtime_requirements_mut: &mut RuntimeGlobals,
+) -> Result<Option<()>> {
+  let chunk = compilation.chunk_by_ukey.expect_get(chunk_ukey);
+
+  // Skip build time chunks
+  if chunk.name() == Some("build time chunk") {
+    return Ok(None);
+  }
+
+  // Only inject EmbedFederationRuntimeModule into runtime chunks
+  let has_runtime = chunk.has_runtime(&compilation.chunk_group_by_ukey);
+  if has_runtime {
+    // Collect federation dependencies snapshot
+    let collected_ids_snapshot = self
```

Comment:
hooks work a little different, since i cannot nest hooks easily, i am passing the tap results from the parent